Lib/importlib/_bootstrap.py

```diff
@@ -424,6 +424,45 @@ def __exit__(self, *args, **kwargs):
         self._lock.release()
 
 
+class _HierarchicalLockManager:
+    """Manages acquisition of multiple module locks in hierarchical order.
+
+    This prevents deadlocks by ensuring all threads acquire locks in the
+    same order (parent modules before child modules).
+    """
+
+    def __init__(self, name):
+        self._name = name
+        self._module_chain = self._get_module_chain(name)
+        self._locks = []
+
+    def _get_module_chain(self, name):
+        """Get all modules in the hierarchy from root to leaf.
+
+        For example: 'a.b.c' -> ['a', 'a.b', 'a.b.c']
+        """
+        parts = name.split('.')
+        return ['.'.join(parts[:i+1]) for i in range(len(parts))]
+
+    def __enter__(self):
+        # Acquire locks for all modules in hierarchy order (parent to child)
+        for module_name in self._module_chain:
+            # Only acquire lock if module is not already fully loaded
+            module = sys.modules.get(module_name)
+            if (module is None or
+                getattr(getattr(module, "__spec__", None), "_initializing", False)):
```

Comment:
is this the right way to do this?

Reply:
In `_get_module_lock` a sentinel object is the default for `sys.modules.get()`, so that might be better. Though I think using None as a sentinel should be fine?